farming/logics/traits/master_chef/farming.rs

```diff
@@ -0,0 +1,482 @@
+        }
+        self.data::<Data>().pool_info.insert(
+            &pool_length,
+            &Pool {
+                acc_arsw_per_share: 0,
+                last_reward_block: Self::env().block_number(),
+                alloc_point,
+            },
+        );
+        self.data::<Data>().pool_info_length = pool_length
+            .checked_add(1)
+            .ok_or(FarmingError::AddOverflow2)?;
+
+        self._emit_log_pool_addition_event(pool_length, alloc_point, lp_token, rewarder);
+        Ok(())
+    }
+
+    #[ink(message)]
+    #[modifiers(only_owner)]
+    fn set(
+        &mut self,
+        pool_id: u32,
+        alloc_point: u32,
+        rewarder: Option<AccountId>,
+        overwrite: bool,
+    ) -> Result<(), FarmingError> {
+        let pool_info = self
+            .get_pool_infos(pool_id)
+            .ok_or(FarmingError::PoolNotFound5)?;
+        self._update_all_pools()?;
+        self.data::<Data>().total_alloc_point = self
+            .data::<Data>()
+            .total_alloc_point
+            .checked_sub(pool_info.alloc_point)
+            .ok_or(FarmingError::SubUnderflow7)?
+            .checked_add(alloc_point)
+            .ok_or(FarmingError::AddOverflow9)?;
+
+        self.data::<Data>().pool_info.insert(
+            &pool_id,
+            &Pool {
+                alloc_point,
+                ..pool_info
+            },
+        );
+        let mut rewarder = rewarder;
+        if overwrite {
+            match rewarder {
+                Some(rewarder_address) => {
+                    self.data::<Data>()
+                        .rewarders
+                        .insert(&pool_id, &rewarder_address)
+                }
+                None => self.data::<Data>().rewarders.remove(&pool_id),
+            }
+        } else {
+            rewarder = self.get_rewarder(pool_id);
+        }
+        self._emit_log_set_pool_event(pool_id, alloc_point, rewarder, overwrite);
+        Ok(())
+    }
+
+    #[ink(message)]
+    fn pending_arsw(&mut self, pool_id: u32, user: AccountId) -> Result<Balance, FarmingError> {
+        let pool = self
+            .get_pool_infos(pool_id)
+            .ok_or(FarmingError::PoolNotFound4)?;
+        let user_info = self
+            .get_user_info(pool_id, user)
+            .ok_or(FarmingError::UserNotFound)?;
+        let mut acc_arsw_per_share = pool.acc_arsw_per_share;
+
+        let lp_supply = self._get_lp_supply(pool_id)?;
+        let current_block = Self::env().block_number();
+
+        if current_block > pool.last_reward_block && lp_supply != 0 {
+            let additional_acc_arsw_per_share =
+                self._calculate_additional_acc_arsw_per_share(&pool, current_block, lp_supply)?;
+            acc_arsw_per_share = acc_arsw_per_share
+                .checked_add(additional_acc_arsw_per_share)
+                .ok_or(FarmingError::AddOverflow8)?;
+        }
+
+        let pending = <u128 as TryInto<i128>>::try_into(
```

Comment:
Using wrapping_add combined with casting might be better, something like 
```rust
    let mut x: u16 = 100;
    let y: i8 = -1;
    x = x.wrapping_add(y as u16);
```

Reply:
Why wrapping_add as there is no addition ? 
```
        pending = int256(
            userInfo.amount.mul(accARSWPerShare) / ACC_ARSW_PRECISION
        ).sub(userInfo.rewardDebt).toUInt256();
```

Reply:
Note that this is possible only under an invariant, the resulting value (here `pending`) is always positive, which holds in our case.

Reply:
Updated